pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java

```diff
@@ -0,0 +1,504 @@
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapred.JobContext;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.MRJobConfig;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
+import org.apache.hadoop.mapreduce.lib.output.LazyOutputFormat;
+import org.apache.pinot.common.Utils;
+import org.apache.pinot.common.config.ColumnPartitionConfig;
+import org.apache.pinot.common.config.IndexingConfig;
+import org.apache.pinot.common.config.SegmentPartitionConfig;
+import org.apache.pinot.common.config.TableConfig;
+import org.apache.pinot.hadoop.io.CombineAvroKeyInputFormat;
+import org.apache.pinot.hadoop.job.mapper.SegmentPreprocessingMapper;
+import org.apache.pinot.hadoop.job.partitioners.GenericPartitioner;
+import org.apache.pinot.hadoop.job.reducer.SegmentPreprocessingReducer;
+import org.apache.pinot.hadoop.utils.PushLocation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.hadoop.job.JobConfigConstants.*;
+
+
+public class SegmentPreprocessingJob extends Configured {
+  private static final Logger _logger = LoggerFactory.getLogger(SegmentPreprocessingJob.class);
+  private final Properties _properties;
+  private Configuration _configuration;
+
+  private boolean _enablePartitioning = false;
+  private boolean _enableSorting = false;
+  private boolean _enableResizing = false;
+
+  private String _partitionColumn;
+  private int _numberOfPartitions;
+  private String _partitionFunction;
+  private String _sortedColumn;
+  private int _numberOfOutputFiles;
+
+  private final String _inputSegmentDir;
+  private final String _preprocessedOutputDir;
+  protected final String _rawTableName;
+
+  private final String _hdfsDefaultClasspathDir;
+  private final String _defaultPermissionsMask;
+  protected final List<PushLocation> _pushLocations;
+
+  private TableConfig _tableConfig;
+  protected FileSystem _fileSystem;
+
+  public SegmentPreprocessingJob(final Properties properties) {
+    _properties = properties;
+    _configuration = new Configuration();
+    setConf(_configuration);
+    _configuration.set("mapreduce.job.user.classpath.first", "true");
+    Utils.logVersions();
+    logProperties();
+
+    if (_properties.getProperty(ENABLE_PARTITIONING) != null) {
+      _enablePartitioning = Boolean.parseBoolean(_properties.getProperty(ENABLE_PARTITIONING));
+    }
+
+    if (_properties.getProperty(ENABLE_SORTING) != null) {
+      _enableSorting = Boolean.parseBoolean(_properties.getProperty(ENABLE_SORTING));
+    }
+
+    if (!_enablePartitioning && _properties.getProperty(ENABLE_RESIZING) != null) {
+      _enableResizing = Boolean.parseBoolean(_properties.getProperty(ENABLE_RESIZING));
+    }
+
```

Comment:
Is this section already in opensource? I know it's part of the pbnj push job. If so, can we merge the logic?

Reply:
Which section do you refer to? If you refer to checking `_enablePartitioning` and `_enableSorting`, it's not in open source yet. The reason why we need this in this job is that it contains 3 components/features, whereas other existing hadoop job like segment creation job is a single feature.

Reply:
the whoel section to read parameters is what i meant